lib/src/http_body_impl.dart

```diff
@@ -19,118 +19,96 @@ class HttpBodyHandlerTransformer
       var charset = contentType.charset;
       if (charset != null) encoding = Encoding.getByName(charset);
       encoding ??= defaultEncoding;
-      return encoding.decoder
+      var buffer = await encoding.decoder
           .bind(stream)
-          .fold(StringBuffer(), (buffer, data) => buffer..write(data))
-          .then((buffer) => _HttpBody('text', buffer.toString()));
+          .fold(StringBuffer(), (buffer, data) => buffer..write(data));
+      return _HttpBody('text', buffer.toString());
     }
 
-    Future<HttpBody> asFormData() {
-      return MimeMultipartTransformer(contentType.parameters['boundary'])
-          .bind(stream)
-          .map((part) => HttpMultipartFormData.parse(part,
-              defaultEncoding: defaultEncoding))
-          .map((multipart) {
-            Future future;
-            if (multipart.isText) {
-              future = multipart
-                  .fold(StringBuffer(), (b, s) => b..write(s))
-                  .then((b) => b.toString());
-            } else {
-              future = multipart
-                  .fold(BytesBuilder(), (b, d) => b..add(d))
-                  .then((b) => b.takeBytes());
-            }
-            return future.then((data) {
-              var filename =
-                  multipart.contentDisposition.parameters['filename'];
-              if (filename != null) {
-                data =
-                    _HttpBodyFileUpload(multipart.contentType, filename, data);
-              }
-              return [multipart.contentDisposition.parameters['name'], data];
-            });
-          })
-          .fold(<Future>[], (List<Future> l, f) => l..add(f))
-          .then((values) => Future.wait(values))
-          .then((parts) {
-            var map = <String, dynamic>{};
-            for (var part in parts) {
-              map[part[0] as String] = part[1]; // Override existing entries.
-            }
-            return _HttpBody('form', map);
-          });
+    Future<HttpBody> asFormData() async {
+      var values =
+          await MimeMultipartTransformer(contentType.parameters['boundary'])
+              .bind(stream)
+              .map((part) => HttpMultipartFormData.parse(part,
+                  defaultEncoding: defaultEncoding))
+              .map((multipart) async {
+        dynamic data;
+        if (multipart.isText) {
+          var buffer = await multipart.fold<StringBuffer>(
+              StringBuffer(), (b, s) => b..write(s));
+          data = buffer.toString();
+        } else {
+          var buffer = await multipart.fold<BytesBuilder>(
+              BytesBuilder(), (b, d) => b..add(d as List<int>));
+          data = buffer.takeBytes();
+        }
+        var filename = multipart.contentDisposition.parameters['filename'];
+        if (filename != null) {
+          data = _HttpBodyFileUpload(multipart.contentType, filename, data);
+        }
+        return [multipart.contentDisposition.parameters['name'], data];
+      }).toList();
```

Comment:
nit: the `toList()` here is not required

Reply:
This is `Stream.toList()` not `Iterable.toList()` so I think it is required.